Add variants to petsc

glennpj opened the pull request:
This PR adds the follolwing variants to the petsc package

- gmp
- jpeg
- libpng
- giflib
- mpfr
- netcdf
- pnetcdf (parallel-netcdf)
- moab
- eigen
- random123
- exodusii
- mstk
- cgns
- memkind
- muparser
- p4est
- saws
- libyaml
- zstd

glennpj committed:
Add variants to petsc

This PR adds the follolwing variants to the petsc package

- gmp
- jpeg
- libpng
- giflib
- mpfr
- netcdf
- pnetcdf (parallel-netcdf)
- moab
- eigen
- random123
- exodusii
- mstk
- cgns
- memkind
- muparser
- p4est
- saws
- libyaml
- zstd

glennpj committed:
Fix flake8 errors

adamjstewart commented:
@balay @barrysmith @jedbrown can you review this?

balay commented on var/spack/repos/builtin/packages/petsc/package.py:
```diff
@@ -105,6 +105,44 @@ class Petsc(Package):
     variant('valgrind', default=False,
             description='Enable Valgrind Client Request mechanism')
+    variant('gmp', default=False,
+            description='Enable GMP support')
```
We'll have to trim this list as some of these packages are primarily dependencies of dependencies.

For ex: zlib is a dependency of hdf5 - so its not listed as a variant - but as:
```
    depends_on('zlib', when='+hdf5')
```
In this case gmp is a dependency of mpfr  [which hardly gets tested].

BTW: Are these changes tested - i.e do spack builds work with these variants enabled?

glennpj replied:
It is possible to have petsc build with gmp and not with mpfr, however, you are correct, building with mpfr requires gmp to be enabled. That is not reflected in the initial PR so would need to be corrected. I did not add a zlib variant for two reasons,
1. It is already the "default" because `+hdf5` is set by default.
1. It may be that zlib should be enabled by default and set as a dependency, but I had not explored that.

With regard to testing, our site's non-spack built petsc-3.10.3 was built including the following variants in this PR:
- netcdf
- mpfr
- libpng
- libjpeg
- gmp

My initial motivation for this PR was to add netcdf support. I then added the other things that I built previous versions of petsc with, and finally added others while I was in there. If any of those seem unneeded/unuseful/silly then I can pull them out. My main desire is for netcdf support, and to a lesser degree, parallel-netcdf.

jedbrown replied:
I think you've missed @balay's point. PETSc does not use GMP. PETSc's configure knows about it so it can find/build it as a dependency for MPFR (which is used in one rarely-used function).

balay replied:
Note: the premise here is: we need variants only for packages that PETSc directly interfaces with.

The reference to hf5, zlib is an example of the case of a dependency of a dependency [as currently implemented in spack]

Also note: PETSc does not have an interface to netcdf  parallel-netcdf. They are dependencies for other pacakges. (i.e a varaint should not be added for them)

* pnetcdf is a dependency for exodusii, netcdf (optional)
* netcdf is a dependency for ADIOS (optional),moab (optional),Trilinos (alternate way of installing exodus,ML etc)

And generally - with shared libraires - one doesn't need to worry about dependencies of dependencies. But in static builds we do - hence this complexity in petsc configure. [apart from the fact that petsc configure can build these dependencies on its own]

glennpj replied:
Well, petsc can build with and link gmp independently of mpfr. Whether it uses it is another story, but,
```
--with-gmp=1 --with-mpfr=0
```
will produce a petsc library linked with libgmp. In any event, if people think it is not a necessary variant, it can be removed.

jedbrown replied:
I would call this a bug in PETSc configure that isn't worth "fixing", except perhaps to convince people who want to turn "everything" on. It's an artifact of PETSc's configure being a sort of package manager. It's easy to check that Satish and I aren't lying to you about PETSc not using GMP. If you add this as a Spack variant, it would only serve to mislead more people.

balay replied:
To illustrate:
```
balay@sb /home/balay/junk
$ cat sz.c
#include <stdio.h>
int main()
{
printf("test\n");
return 0;
}
balay@sb /home/balay/junk
$ gcc sz.c
balay@sb /home/balay/junk
$ ldd a.out
	linux-vdso.so.1 (0x00007ffcae5ef000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f683e203000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f683e3f5000)
balay@sb /home/balay/junk
$ mpicc sz.c
balay@sb /home/balay/junk
$ ldd a.out
	linux-vdso.so.1 (0x00007ffcfd7e8000)
	libmpi.so.12 => /home/balay/soft/mpich-3.3.2/lib/libmpi.so.12 (0x00007f9e199af000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f9e197bf000)
	libgfortran.so.5 => /lib64/libgfortran.so.5 (0x00007f9e194f7000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f9e193b1000)
	libOpenCL.so.1 => /lib64/libOpenCL.so.1 (0x00007f9e1938f000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f9e1936d000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f9e19360000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f9e19345000)
	libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007f9e192fb000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9e19ef2000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f9e192f4000)
```

balay replied:
i.e just because `libmpi.so.12` is linked in does not imply the above test code is dependent on MPI

glennpj replied:
I believe you, and always did, I was just presenting what I was observing. It sounds to me like gmp and mpfr variants can be nixed.

balay replied:
```
$ git grep PETSC_HAVE_MPFR
src/dm/dt/interface/dt.c:#if defined(PETSC_HAVE_MPFR)
src/dm/dt/interface/dt.c:#if defined(PETSC_HAVE_MPFR)
src/dm/dt/tests/ex3.c:#if defined(PETSC_HAVE_MPFR)
```
gmp should be nixed. mpfr can be kep [however configure will insist on passing --with-gmp option for -with-mpfr to work]

netcdf, pnetcdf variants should be nixed aswell.

Will check the complete list...

glennpj replied:
I just saw the most recent comment from @balay. I am somewhat confused by the following:
>Note: the premise here is: we need variants only for packages that PETSc directly interfaces with.
...
Also note: PETSc does have an interface to netcdf parallel-netcdf. They are dependencies for other pacakges. (i.e a varaint should not be added for them)

The first statement indicates that a variant would be viable if there is an interface but the second statement seems to contradict that. Or, did you mean that, "PETSc does **not** have an interface to netcdf parallel-netcdf"?

Sorry, I seem to be learning that I know even less about PETSc than I thought.

balay replied:
Sorry - typo. `petsc does not have an interface to netcdf, parallel-netcdf`

balay replied:
* gmp     - nixed
* jpeg
* libpng   (configure checks for zlib as dependencies)
* giflib
* mpfr	   (configure checks for gmp as dependencies)
* netcdf  - nixed
* pnetcdf - nixed
* moab
* eigen   - nixed
* random123
* exodusii (configure checks for hdf5, netcdf, pnetcdf as dependencies)
* mstk    - nixed
* cgns     (configure checks for hdf5 as dependencies)
* memkind
* muparser - nixed (don't know why we have this)
* p4est    (configure checks for zlib as dependencies)
* saws
* libyaml
* zstd	   - nixed (don't know why we have this)

glennpj replied:
> i.e just because `libmpi.so.12` is linked in does not imply the above test code is dependent on MPI

Yes, I understand that. I was trusting the configuration mechanism to use what it was looking for, which was apparently, a mistake.

balay replied:
Basically configure is not rejecting mpicc when user specifies it [even-though its not required in this specific case]. And it trusts the user need it for other reasons.. [and checks if things work with it or not]

jedbrown replied:
`zstd` can be dropped here. I've been using it in a branch with a student, but that feature hasn't landed yet.

glennpj replied:
So, the missing piece for me was not understanding there was a distinction between packages presented in `configure --help` that interfaced with PETSc and their dependencies. It would be nice if that could be made more clear in the `configure --help` output so that people who are not intimately familiar with PETSc, but just trying to package it for their users, would have more guidance.

The question that I have at this point is should I pursue adding the options from the list that @balay curated to the spack package or just drop this PR?

jedbrown replied:
The information is in the configure files; see `self.useddirectly = 0` for those that are only indirect dependencies.
```console
$ git grep useddirecly
```
Yes, someone could make that visible in the `./configure --help` output. Preferably not more lines; the output is long enough already.

glennpj committed:
Additional changes to Petsc recipe

This commit addresses the issues with dependencies that were brought up
in the comments. There are also a few other enhancements.

- the language of the new variant descriptions was changed to be more
  consistent with what was already in the recipe
- an explicit '+mpi' was added to the depends_on('hypre...') directives
- an explicit '+mpi' was added to the depends_on('trilinos...')
  directives
- the run time error checking for '~mpi' was replaced with 'conflicts()'
  directives that will cause the install to fail sooner
- additional variants that were 'parallel only' were added to the '~mpi'
  check

glennpj committed:
Set the '~mpi`' conflicts msg to a variable

glennpj commented:
Restart tests.

glennpj commented:
@adamjstewart When you get a chance could you take a look at this PR to see if I addressed everything.

adamjstewart reviewed (approved):
Looks fine to me, but I don't know anything about `petsc`. Waiting for one of the maintainers to approve before merging.

glennpj commented:
@jedbrown and/or @balay Could you take a look at this to see if everything has been addressed?

Thanks.

balay commented on var/spack/repos/builtin/packages/petsc/package.py:
```diff
@@ -105,10 +105,47 @@ class Petsc(Package):
+    conflicts('+mumps', when='~mpi', msg=mpi_msg)
+    conflicts('+p4est', when='~mpi', msg=mpi_msg)
+    conflicts('+superlu-dist', when='~mpi', msg=mpi_msg)
+    conflicts('+trilinos', when='~mpi', msg=mpi_msg)
```
This is tricky.

petsc can build with hdf5\~mpi, mumps\~mpi.

And the other packages (exodusii cgns) have dependencies on hdf5 [its not clear if we tested them with hdf5\~mpi]. So I'm not sure what's appropriate here.

I see this issue is with current code in spack - and not related to this MR. [its just moving the check]. So I guess its fine to keep these conflicts for now - and update once we can test those combinations in spack

balay commented:
@glennpj thanks for the improvements! sorry about the delay in response

glennpj commented:
> @glennpj thanks for the improvements! sorry about the delay in response

No worries. I learned quite a bit working on this PR and that is always a good thing.
